cache/grpcproxy/grpcproxy.go

```diff
@@ -0,0 +1,287 @@
+package grpcproxy
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/buchgr/bazel-remote/v2/cache"
+	"github.com/buchgr/bazel-remote/v2/utils/backendproxy"
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	asset "github.com/buchgr/bazel-remote/v2/genproto/build/bazel/remote/asset/v1"
+	pb "github.com/buchgr/bazel-remote/v2/genproto/build/bazel/remote/execution/v2"
+	bs "google.golang.org/genproto/googleapis/bytestream"
+)
+
+type GrpcClients struct {
+	asset asset.FetchClient
+	bs    bs.ByteStreamClient
+	ac    pb.ActionCacheClient
+	cas   pb.ContentAddressableStorageClient
+}
+
+func NewGrpcClients(cc *grpc.ClientConn) *GrpcClients {
+	return &GrpcClients{
+		asset: asset.NewFetchClient(cc),
+		bs:    bs.NewByteStreamClient(cc),
+		ac:    pb.NewActionCacheClient(cc),
+		cas:   pb.NewContentAddressableStorageClient(cc),
+	}
+}
+
+type remoteGrpcProxyCache struct {
+	clients      *GrpcClients
+	uploadQueue  chan<- backendproxy.UploadReq
+	accessLogger cache.Logger
+	errorLogger  cache.Logger
+}
+
+func New(clients *GrpcClients,
+	accessLogger cache.Logger, errorLogger cache.Logger,
+	numUploaders, maxQueuedUploads int) (cache.Proxy, error) {
+
+	proxy := &remoteGrpcProxyCache{
+		clients:      clients,
+		accessLogger: accessLogger,
+		errorLogger:  errorLogger,
+	}
+
+	proxy.uploadQueue = backendproxy.StartUploaders(proxy, numUploaders, maxQueuedUploads)
+
+	return proxy, nil
+}
+
+// Helper function for logging responses
+func logResponse(logger cache.Logger, method string, msg string, resource string) {
+	logger.Printf("GRPC PROXY %s %s: %s", method, msg, resource)
+}
+
+func (r *remoteGrpcProxyCache) UploadFile(item backendproxy.UploadReq) {
+	switch item.Kind {
+	case cache.RAW:
+		fallthrough
```

Comment:
I think this case deserves a comment, and we can maybe upgrade that to a log message later.